Feature/ruby enumerables project

majovanilla opened the pull request:
This is the request for the code review of the solo project Ruby Enumerables

- [x] Create all methods required.
- [x] Run tests with arrays and ranges.
- [x] Change stickler errors.

majovanilla committed:
ADD STICKLER FILES

majovanilla committed:
CREATE MY_EACH METHOD

majovanilla committed:
ADD METHODS WITH BLOCK_GIVEN? VALIDATION

my_each, my_each_with_index, my_select, my_all and my_any methods

majovanilla committed:
ALL METHODS CREATED

Add all methods and tested them with and without a block given

majovanilla committed:
EDIT MY_MAP TO TAKE A PROC

majovanilla committed:
EDIT METHODS TO TAKE IN RANGES

majovanilla committed:
ALL METHODS CHECKED

majovanilla committed:
EDIT STICKLER ERRORS

maelfosso commented on enumerables.rb:
```diff
@@ -0,0 +1,118 @@
+      all = false unless i
+    end
+    all
+  end
```
```ruby
irb(main):009:0> true_array = [1, true, 'hi', []]
=> [1, true, "hi", []]
irb(main):010:0> true_array.all?
=> true
irb(main):011:0> true_array.my_all?
Traceback (most recent call last):
        5: from /home/elvisfosso/.rbenv/versions/2.5.1/bin/irb:11:in `<main>'
        4: from (irb):11
        3: from enumerables.rb:43:in `my_all?'
        2: from enumerables.rb:10:in `my_each'
        1: from enumerables.rb:44:in `block in my_all?'
LocalJumpError (no block given (yield))
irb(main):012:0> 
```

maelfosso replied:
```ruby
irb(main):013:0> array
=> [2, 8, 5, 0, 5, 6, 5, 0, 0, 5, 7, 6, 0, 4, 6, 7, 1, 3, 6, 2, 0, 3, 1, 6, 8, 0, 8, 3, 1, 7, 3, 0, 2, 5, 8, 5, 6, 3, 3, 2, 1, 1, 4, 1, 6, 6, 0, 3, 3, 7, 4, 5, 1, 1, 5, 0, 7, 6, 2, 2, 1, 4, 8, 3, 7, 3, 7, 8, 5, 8, 3, 7, 5, 4, 0, 6, 8, 7, 6, 7, 1, 6, 0, 7, 0, 1, 3, 1, 6, 6, 5, 3, 2, 4, 1, 3, 1, 2, 3, 1]
irb(main):014:0> array.all?(Integer)
=> true
irb(main):015:0> array.my_all?(Integer)
Traceback (most recent call last):
        3: from /home/elvisfosso/.rbenv/versions/2.5.1/bin/irb:11:in `<main>'
        2: from (irb):15
        1: from enumerables.rb:41:in `my_all?'
ArgumentError (wrong number of arguments (given 1, expected 0))
irb(main):016:0> 
```

maelfosso replied:
```ruby
irb(main):017:0> words
=> ["dog", "door", "rod", "blade"]
irb(main):018:0> words.all?(/d/)
=> true
irb(main):019:0> words.my_all?(/d/)
Traceback (most recent call last):
        3: from /home/elvisfosso/.rbenv/versions/2.5.1/bin/irb:11:in `<main>'
        2: from (irb):19
        1: from enumerables.rb:41:in `my_all?'
ArgumentError (wrong number of arguments (given 1, expected 0))
irb(main):020:0> 
```

maelfosso replied:
```ruby
irb(main):023:0> array
=> [2, 8, 5, 0, 5, 6, 5, 0, 0, 5, 7, 6, 0, 4, 6, 7, 1, 3, 6, 2, 0, 3, 1, 6, 8, 0, 8, 3, 1, 7, 3, 0, 2, 5, 8, 5, 6, 3, 3, 2, 1, 1, 4, 1, 6, 6, 0, 3, 3, 7, 4, 5, 1, 1, 5, 0, 7, 6, 2, 2, 1, 4, 8, 3, 7, 3, 7, 8, 5, 8, 3, 7, 5, 4, 0, 6, 8, 7, 6, 7, 1, 6, 0, 7, 0, 1, 3, 1, 6, 6, 5, 3, 2, 4, 1, 3, 1, 2, 3, 1]
irb(main):024:0> array.all?(3)
=> false
irb(main):025:0> array.my_all?(3)
Traceback (most recent call last):
        3: from /home/elvisfosso/.rbenv/versions/2.5.1/bin/irb:11:in `<main>'
        2: from (irb):25
        1: from enumerables.rb:41:in `my_all?'
ArgumentError (wrong number of arguments (given 1, expected 0))
irb(main):026:0> 
```

maelfosso commented on enumerables.rb:
```diff
@@ -0,0 +1,118 @@
+      any = true unless block_given? && i
+    end
+    any
+  end
```
```ruby
irb(main):028:0> false_array = [nil, false, nil, false]
=> [nil, false, nil, false]
irb(main):029:0> false_array.any?
=> false
irb(main):030:0> false_array.my_any?
=> true
irb(main):031:0>
```

maelfosso replied:
```ruby
irb(main):033:0> array
=> [2, 8, 5, 0, 5, 6, 5, 0, 0, 5, 7, 6, 0, 4, 6, 7, 1, 3, 6, 2, 0, 3, 1, 6, 8, 0, 8, 3, 1, 7, 3, 0, 2, 5, 8, 5, 6, 3, 3, 2, 1, 1, 4, 1, 6, 6, 0, 3, 3, 7, 4, 5, 1, 1, 5, 0, 7, 6, 2, 2, 1, 4, 8, 3, 7, 3, 7, 8, 5, 8, 3, 7, 5, 4, 0, 6, 8, 7, 6, 7, 1, 6, 0, 7, 0, 1, 3, 1, 6, 6, 5, 3, 2, 4, 1, 3, 1, 2, 3, 1]
irb(main):034:0> array.any?(Integer)
=> true
irb(main):035:0> array.my_any?(Integer)
Traceback (most recent call last):
        3: from /home/elvisfosso/.rbenv/versions/2.5.1/bin/irb:11:in `<main>'
        2: from (irb):35
        1: from enumerables.rb:52:in `my_any?'
ArgumentError (wrong number of arguments (given 1, expected 0))
irb(main):036:0> 
```

maelfosso replied:
```ruby
irb(main):038:0> words
=> ["dog", "door", "rod", "blade"]
irb(main):039:0> words.any?(/z/)
=> false
irb(main):040:0> words.my_any?(/z/)
Traceback (most recent call last):
        3: from /home/elvisfosso/.rbenv/versions/2.5.1/bin/irb:11:in `<main>'
        2: from (irb):40
        1: from enumerables.rb:52:in `my_any?'
ArgumentError (wrong number of arguments (given 1, expected 0))
irb(main):041:0> 
```

maelfosso replied:
```ruby
irb(main):043:0> words
=> ["dog", "door", "rod", "blade"]
irb(main):044:0> words[0] = 'cat'
=> "cat"
irb(main):045:0> words.any?('cat')
=> true
irb(main):046:0> words.my_any?('cat')
Traceback (most recent call last):
        3: from /home/elvisfosso/.rbenv/versions/2.5.1/bin/irb:11:in `<main>'
        2: from (irb):46
        1: from enumerables.rb:52:in `my_any?'
ArgumentError (wrong number of arguments (given 1, expected 0))
irb(main):047:0> 
```

maelfosso commented on enumerables.rb:
```diff
@@ -0,0 +1,118 @@
+      end
+    end
+    none
+  end
```
```ruby
irb(main):062:0> array
=> [2, 8, 5, 0, 5, 6, 5, 0, 0, 5, 7, 6, 0, 4, 6, 7, 1, 3, 6, 2, 0, 3, 1, 6, 8, 0, 8, 3, 1, 7, 3, 0, 2, 5, 8, 5, 6, 3, 3, 2, 1, 1, 4, 1, 6, 6, 0, 3, 3, 7, 4, 5, 1, 1, 5, 0, 7, 6, 2, 2, 1, 4, 8, 3, 7, 3, 7, 8, 5, 8, 3, 7, 5, 4, 0, 6, 8, 7, 6, 7, 1, 6, 0, 7, 0, 1, 3, 1, 6, 6, 5, 3, 2, 4, 1, 3, 1, 2, 3, 1]
irb(main):063:0> array.none?(String)
=> true
irb(main):064:0> array.my_none?(String)
Traceback (most recent call last):
        3: from /home/elvisfosso/.rbenv/versions/2.5.1/bin/irb:11:in `<main>'
        2: from (irb):64
        1: from enumerables.rb:62:in `my_none?'
ArgumentError (wrong number of arguments (given 1, expected 0))
irb(main):065:0> 
```

maelfosso replied:
```ruby
irb(main):067:0> words
=> ["cat", "door", "rod", "blade"]
irb(main):068:0> words.none?(/z/)
=> true
irb(main):069:0> words.my_none?(/z/)
Traceback (most recent call last):
        3: from /home/elvisfosso/.rbenv/versions/2.5.1/bin/irb:11:in `<main>'
        2: from (irb):69
        1: from enumerables.rb:62:in `my_none?'
ArgumentError (wrong number of arguments (given 1, expected 0))
irb(main):070:0> 
```

maelfosso replied:
```ruby
irb(main):072:0> words
=> ["cat", "door", "rod", "blade"]
irb(main):073:0> words[0] = 5
=> 5
irb(main):074:0> words.none?(5)
=> false
irb(main):075:0> words.my_none?(5)
Traceback (most recent call last):
        3: from /home/elvisfosso/.rbenv/versions/2.5.1/bin/irb:11:in `<main>'
        2: from (irb):75
        1: from enumerables.rb:62:in `my_none?'
ArgumentError (wrong number of arguments (given 1, expected 0))
irb(main):076:0> 
```

maelfosso commented on enumerables.rb:
```diff
@@ -0,0 +1,118 @@
+      counter += 1 if block_given? && yield(i)
+    end
+    counter
+  end
```
```ruby
irb(main):078:0> array
=> [2, 8, 5, 0, 5, 6, 5, 0, 0, 5, 7, 6, 0, 4, 6, 7, 1, 3, 6, 2, 0, 3, 1, 6, 8, 0, 8, 3, 1, 7, 3, 0, 2, 5, 8, 5, 6, 3, 3, 2, 1, 1, 4, 1, 6, 6, 0, 3, 3, 7, 4, 5, 1, 1, 5, 0, 7, 6, 2, 2, 1, 4, 8, 3, 7, 3, 7, 8, 5, 8, 3, 7, 5, 4, 0, 6, 8, 7, 6, 7, 1, 6, 0, 7, 0, 1, 3, 1, 6, 6, 5, 3, 2, 4, 1, 3, 1, 2, 3, 1]
irb(main):079:0> array.count(3)
=> 15
irb(main):080:0> array.my_count(3)
Traceback (most recent call last):
        3: from /home/elvisfosso/.rbenv/versions/2.5.1/bin/irb:11:in `<main>'
        2: from (irb):80
        1: from enumerables.rb:77:in `my_count'
ArgumentError (wrong number of arguments (given 1, expected 0))
irb(main):081:0> 
```

maelfosso commented on enumerables.rb:
```diff
@@ -0,0 +1,118 @@
+      end
+    end
+    memo
+  end
```
```ruby
irb(main):083:0> range = Range.new(5, 50)
=> 5..50
irb(main):084:0> range.inject(4) { |prod, n| prod * n }
=> 5069015533618896340602101361010794807396273594826752000000000000
irb(main):085:0> range.my_inject(4) { |prod, n| prod * n }
Traceback (most recent call last):
        3: from /home/elvisfosso/.rbenv/versions/2.5.1/bin/irb:11:in `<main>'
        2: from (irb):85
        1: from enumerables.rb:97:in `my_inject'
ArgumentError (wrong number of arguments (given 1, expected 0))
irb(main):086:0> 
```

maelfosso replied:
```ruby
irb(main):088:0> array
=> [2, 8, 5, 0, 5, 6, 5, 0, 0, 5, 7, 6, 0, 4, 6, 7, 1, 3, 6, 2, 0, 3, 1, 6, 8, 0, 8, 3, 1, 7, 3, 0, 2, 5, 8, 5, 6, 3, 3, 2, 1, 1, 4, 1, 6, 6, 0, 3, 3, 7, 4, 5, 1, 1, 5, 0, 7, 6, 2, 2, 1, 4, 8, 3, 7, 3, 7, 8, 5, 8, 3, 7, 5, 4, 0, 6, 8, 7, 6, 7, 1, 6, 0, 7, 0, 1, 3, 1, 6, 6, 5, 3, 2, 4, 1, 3, 1, 2, 3, 1]
irb(main):089:0> array.inject(:+)
=> 380
irb(main):090:0> array.my_inject(:+)
Traceback (most recent call last):
        3: from /home/elvisfosso/.rbenv/versions/2.5.1/bin/irb:11:in `<main>'
        2: from (irb):90
        1: from enumerables.rb:97:in `my_inject'
ArgumentError (wrong number of arguments (given 1, expected 0))
irb(main):091:0>

maelfosso reviewed (changes requested):
# Not good to be approved :disappointed: 

Hello @majovainilla 

For a first step, it's very very awesome :clap: 
However, there is a lot of bug to fix. Please, read the comments below to know more about them and then kindly make the changes and submit a new code review request.

Thanks,
**Mael FOSSO**
[Github](https://github.com) | [Twitter](https://twitter.com) | [LinkedIn](https://www.linkedin.com/in/mael-elvis-fosso-650b6346/)

majovanilla committed:
EDIT README FILE

majovanilla committed:
EDIT METHODS' INPUTS

Add method to take inputs like string, integer or regex

majovanilla committed:
EDIT ERRORS

Add new method to check for different inputs
Edit inject method to improve performance

majovanilla committed:
CHANGE STICKLER SUGGESTIONS

samgaco commented on enumerables.rb:
```diff
@@ -0,0 +1,146 @@
+  def my_each_with_index
+    i = 0
+    while i < size
+      return to_enum if block_given?
```
- [x] my_each_with_index calls the given block once for each element in self

```
irb(main):147:0> [1,2,3].each_with_index do |i,ind| puts i,ind end
1
0
2
1
3
2
=> [1, 2, 3]
irb(main):148:0> [1,2,3].my_each_with_index do |i,ind| puts i,ind end
=> #<Enumerator: [1, 2, 3]:each>
```

- [x] enumerables my_each_with_index returns an enumerator if no block is given

samgaco commented on enumerables.rb:
```diff
@@ -0,0 +1,146 @@
+  def my_inject(start = nil, symbol = nil)
+    memo = 0
+    unless block_given?
+      return to_enum if start.nil? && symbol.nil?
```
- [x] my_inject combines all elements of enum by applying a binary operation, specified by a block

```
irb(main):151:0> [1,2,3].my_inject(:*)
=> 0
irb(main):152:0> [1,2,3].inject(:*)
=> 6
```

samgaco reviewed (changes requested):
HI @majovainilla :high_brightness: 

Some changes are still required.

- [x] Please, fix all the linter errors, the linter checks should be green.

- [x] Address the failures in the comments below, my_each_with_index has two missing functionalities and my_inject is also not 100% correct

Please :pray: work on these changes and submit back for another code review once you are done. Good spirit :muscle: 


----------------------

* ###### STATUS OF THE PROJECT: REQUESTED CHANGES :stop_sign:

majovanilla committed:
ADD COMMENT TO IGNORE COMPLEXITY IN RUBOCOP

majovanilla committed:
EDIT INJECT AND EACH WITH INDEX

Edit inject to correctly use symbol
Edit my_each_with_index to return object at the end

majovanilla committed:
DISABLE COMPLEXITY IN RUBOCOP

majovanilla commented:
Added correct behavior to the errors above. 
Corrected all linter error

majovanilla committed:
EDIT ERRORS ON MY_ALL AND CHECK_INPUT IN REGEXP

samgaco commented on enumerables.rb:
```diff
@@ -0,0 +1,151 @@
+  def my_inject(start = nil, symbol = nil)
+    if block_given?
+      memo = start if start
+      if start.nil?
```
- [x] my_inject when a symbol is specified combines each element of the collection by applying the symbol as a named method


```
irb(main):249:0> ok = [1,2,3,4,5]
=> [1, 2, 3, 4, 5]
irb(main):250:0> ok.my_inject(:+)
=> 15
irb(main):251:0> ok.my_inject(:+)
=> 14
irb(main):252:0> ok.my_inject(:+)
=> 12
irb(main):253:0> ok.my_inject(:+)
=> 9
irb(main):254:0> ok.inject(:+)
=> 5
irb(main):255:0> ok.inject(:+)
=> 5
irb(main):256:0> ok.inject(:+)
=> 5
irb(main):257:0> ok.inject(:+)
=> 5
irb(main):258:0> ok.inject(:+)
=> 5
irb(main):259:0> ok.inject(:+)
=> 5
irb(main):260:0> ok.my_inject(:+)
=> 5
irb(main):261:0> ok.my_inject(:+)
=> nil
irb(main):262:0> ok.my_inject(:+)
=> nil
``` 

- [x] my_inject combines all elements of enum by applying a binary operation, specified by a block


```
irb(main):270:0> arr = [1,2,3,4,5]
=> [1, 2, 3, 4, 5]
irb(main):271:0> arr.inject {|n,j| n+j}
=> 15
irb(main):272:0> arr.inject {|n,j| n+j}
=> 15
irb(main):273:0> arr.inject {|n,j| n+j}
=> 15
irb(main):274:0> arr.inject {|n,j| n+j}
=> 15
irb(main):275:0> arr.inject {|n,j| n+j}
=> 15
irb(main):276:0> arr.inject {|n,j| n+j}
=> 15
irb(main):277:0> arr.my_inject {|n,j| n+j}
=> 15
irb(main):278:0> arr.my_inject {|n,j| n+j}
=> 14
irb(main):279:0> arr.my_inject {|n,j| n+j}
=> 12
irb(main):280:0> arr.my_inject {|n,j| n+j}
=> 9
irb(main):281:0> arr.my_inject {|n,j| n+j}
=> 5
irb(main):282:0> arr.inject {|n,j| n+j}
=> nil
irb(main):283:0> arr.inject {|n,j| n+j}
=> nil

```

majovanilla replied:
Thanks for the review! I was just testing the two cases you told me and I'm getting the right answer. 

`irb(main):002:0> [1, 2, 3, 4, 5].my_inject {|n,j| n+j}`
`=> 15`
`irb(main):003:0> [1, 2, 3, 4, 5].inject {|n,j| n+j}`
`=> 15`
`irb(main):004:0> [1, 2, 3, 4, 5].my_inject(:+)`
`=> 15`
`irb(main):005:0> [1, 2, 3, 4, 5].inject(:+)`
`=> 15`

samgaco replied:
Hi!, try saving the array into a variable and repeating the procedure a few times

samgaco replied:
```
arr = [1,2,3,4,5]
```

and then 

```
irb(main):274:0> arr.inject {|n,j| n+j}
=> 15
irb(main):275:0> arr.inject {|n,j| n+j}
=> 15
irb(main):276:0> arr.inject {|n,j| n+j}
=> 15
irb(main):277:0> arr.my_inject {|n,j| n+j}
=> 15
irb(main):278:0> arr.my_inject {|n,j| n+j}
=> 14
irb(main):279:0> arr.my_inject {|n,j| n+j}
=> 12
irb(main):280:0> arr.my_inject {|n,j| n+j}
=> 9
irb(main):281:0> arr.my_inject {|n,j| n+j}
=> 5
```

samgaco reviewed (changes requested):
Hi @majovainilla :high_brightness: 

**my_inject**  keeps giving some problems, it seems like it is affecting the array it's passed, I left some comments below with some examples. 

Regards.

----------------------------

 * ###### STATUS OF THE PROJECT: REQUESTED CHANGES :stop_sign:

majovanilla committed:
Edit my_inject method to work with symbols

majovanilla committed:
Delete comments in file

majovanilla committed:
Add validations to my_inject to pass test cases

majovanilla committed:
Improve my_inject method to pass all rspec tests

samgaco commented on enumerables.rb:
```diff
@@ -0,0 +1,155 @@
+    new_array = to_a
+    memo = start
+    if start && symbol
+      my_each do |i|
```
still, there seems to be a small error, when starting a range from a number that is not 1, the results differ compared against the original:

```
irb#1(main):032:0> (5..7).inject(4) { |prod, n| prod * n }
=> 840
irb#1(main):033:0> (5..7).my_inject(4) { |prod, n| prod * n }
=> 168
irb#1(main):034:0> (2..7).my_inject(4) { |prod, n| prod * n }
=> 10080
irb#1(main):035:0> (2..7).inject(4) { |prod, n| prod * n }
=> 20160
irb#1(main):036:0> (1..7).inject(4) { |prod, n| prod * n }
=> 20160
irb#1(main):037:0> (1..7).my_inject(4) { |prod, n| prod * n }
=> 20160

```

majovanilla replied:
Thanks Samuel! I've been refactoring that method like crazy to make it work. I just checked what was wrong. Thanks for your reviews!!

samgaco reviewed (changes requested):
Hi @majovainilla :high_brightness: 

I left in the comments one failure that is still remaining, seems to be the last stepping stone before the approval. I appreciate your patience and continuous hard work. Good spirit :+1: 


---------------------------------

* ###### STATUS OF THE PROJECT: REQUESTED CHANGES :stop_sign:

majovanilla committed:
Improve my_inject method to accept start value and block

samgaco commented on enumerables.rb:
```diff
@@ -125,9 +125,7 @@ def my_inject(start = nil, symbol = nil)
-      my_each_with_index do |e, i|
-        next if i.zero?
-
+      my_each do |e|
```
:medal_military:

samgaco reviewed (approved):
excellent job @majovainilla hope you keep up the great work and diligence moving on :muscle: 

You can now merge and submit the project completion form. Regards!

-----------------------------------

* ###### STATUS OF THE PROJECT: APPROVED :heavy_check_mark:
